proof/refine/ARM/SchedContext_R.thy

```diff
@@ -8,8 +8,521 @@ theory SchedContext_R
+  (simp: crunch_simps wp: crunch_wps)
+
+crunches setConsumed
+  for pred_tcb_at'[wp]: "pred_tcb_at' proj P t"
+  (wp: crunch_wps simp: crunch_simps)
+
+lemma setConsumed_invs':
+  "setConsumed scp buffer \<lbrace>invs'\<rbrace>"
+  apply (simp add: setConsumed_def)
+  by (wpsimp wp: schedContextUpdateConsumed_invs'
+      | strengthen tcb_at_invs')+
+
+lemma schedContextCompleteYieldTo_invs'[wp]:
+  "\<lbrace>invs' and sch_act_simple and tcb_at' thread\<rbrace>
+   schedContextCompleteYieldTo thread
+   \<lbrace>\<lambda>_. invs'\<rbrace>"
+  unfolding schedContextCompleteYieldTo_def
+  by (wpsimp wp: schedContextCancelYieldTo_invs' setConsumed_invs'
+                 hoare_drop_imp hoare_vcg_if_lift2
+           simp: sch_act_simple_def)
+
+lemma setConsumed_corres:
+ "corres dc ((\<lambda>s. tcb_at (cur_thread s) s) and case_option \<top> in_user_frame buf and sc_at scp)
+            (\<lambda>s. tcb_at' (ksCurThread s) s \<and> case_option \<top> valid_ipc_buffer_ptr' buf s \<and> sc_at' scp s)
+            (set_consumed scp buf)
+            (setConsumed scp buf)"
+  apply (simp add: set_consumed_def setConsumed_def)
+  apply (rule corres_guard_imp)
+    apply (rule corres_split [OF _ schedContextUpdateConsumed_corres])
+      apply (rule corres_split [OF _ gct_corres], simp)
+        apply (rule corres_split [OF set_mi_corres set_mrs_corres])
+  by (wpsimp wp: hoare_case_option_wp simp: setTimeArg_def)+
+
+lemma get_tcb_yield_to_corres:
+  "corres (=) (pspace_aligned and pspace_distinct and tcb_at t) \<top>
+          (get_tcb_obj_ref tcb_yield_to t) (threadGet tcbYieldTo t)"
+  apply (rule_tac Q="tcb_at' t" in corres_cross_add_guard)
+   apply (fastforce dest!: state_relationD elim!: tcb_at_cross)
+  apply (simp add: get_tcb_obj_ref_def getBoundNotification_def)
+  apply (rule corres_guard_imp)
+    apply (rule threadget_corres)
+    apply (simp add: tcb_relation_def)+
+  done
+
+lemma tcb_yield_to_update_corres:
+  "corres dc (pspace_aligned and pspace_distinct and tcb_at t) \<top>
+          (set_tcb_obj_ref tcb_yield_to_update t yt) (threadSet (tcbYieldTo_update (\<lambda>_. yt)) t)"
+  apply (rule_tac Q="tcb_at' t" in corres_cross_add_guard)
+   apply (fastforce dest!: state_relationD elim!: tcb_at_cross)
+  apply (simp add: set_tcb_obj_ref_def)
+  apply (subst thread_set_def[simplified, symmetric])
+  apply (rule corres_guard_imp)
+    apply (rule threadset_corres, simp_all add: tcb_relation_def)
+  done
+
 context begin interpretation Arch . (*FIXME: arch_split*)
 
+lemma complete_yield_to_corres:
+  "corres dc (invs and tcb_at thread) (invs' and tcb_at' thread)
+    (complete_yield_to thread) (schedContextCompleteYieldTo thread)"
+  apply (simp add: complete_yield_to_def schedContextCompleteYieldTo_def)
+  apply (subst maybeM_when)
+  apply (rule corres_guard_imp)
+    apply (rule corres_split [OF _ get_tcb_yield_to_corres], simp)
+      apply (rule corres_when2[OF refl])
+      apply (clarsimp, wpfix)
```

Comment:
Can we avoid having `wpfix` in our final proof or is this something we can permit?

Reply:
I don't think we can always eliminate it.